File: ProjecTree-Inference/app/services/portfolio_service.py
"""포트폴리오 생성 서비스"""

from app.api.schemas.portfolio import PortfolioGenerateRequest, PortfolioGenerateResponse
from app.agents.portfolio.graph import portfolio_graph
from app.core.log import langfuse_handler
import logging
import traceback

logger = logging.getLogger(__name__)


class PortfolioService:
    """포트폴리오 생성 서비스 클래스
    
    LLM Agent를 활용하여 사용자의 프로젝트 정보를 기반으로 포트폴리오를 생성합니다.
    """
    
    def __init__(self):
        pass
    
    async def generate_portfolio(
        self, 
        request: PortfolioGenerateRequest
    ) -> PortfolioGenerateResponse:
        """포트폴리오 생성
        
        Args:
            request: 포트폴리오 생성 요청 정보
                - project_title: 프로젝트 제목
                - project_description: 프로젝트 설명
                - project_head_count: 프로젝트 인원수
                - user_task_schemas: 사용자가 수행한 작업 목록
        
        Returns:
            생성된 포트폴리오 내용 및 요약
        """
        logger.info(f"[PortfolioService] 포트폴리오 생성 시작 - project_title: {request.project_title}")
        
        try:
            # 요청 데이터를 State 형식으로 변환
            user_tasks = []
            for task in request.user_task_schemas:
                task_dict = {
                    "task_name": task.task_name,
                    "task_description": task.task_description,
                    "task_note": task.task_note,
                    "tech_stack": None
                }
                if task.tech_stack:
                    task_dict["tech_stack"] = {
                        "name": task.tech_stack.name,
                        "advantage": task.tech_stack.advantage,
                        "disadvantage": task.tech_stack.disadvantage,
                        "description": task.tech_stack.description
                    }
                user_tasks.append(task_dict)
            
            # 그래프 실행
            logger.info(f"[PortfolioService] portfolio_graph 호출 시작")
            result = await portfolio_graph.ainvoke(
                {
                    "project_title": request.project_title,
                    "project_description": request.project_description,
                    "project_head_count": request.project_head_count,
                    "project_start_date": request.project_start_date,
                    "project_end_date": request.project_end_date,
                    "project_tech_stack": request.project_tech_stack,
                    "user_tasks": user_tasks,
                    "retry_count": 0
                },
                config={"callbacks": [langfuse_handler]}
            )
            logger.info(f"[PortfolioService] portfolio_graph 호출 완료")
            
            portfolio_content = result.get("portfolio_content", "포트폴리오 생성에 실패했습니다.")
            summary = result.get("summary", f"{request.project_title} 프로젝트 포트폴리오")
            
            logger.info(f"[PortfolioService] 포트폴리오 생성 완료 - project_title: {request.project_title}")
            
            return PortfolioGenerateResponse(
                portfolio_content=portfolio_content,
                summary=summary
            )
            
        except Exception as e:
            logger.error(f"[PortfolioService] 포트폴리오 생성 실패 - project_title: {request.project_title}")
            logger.error(f"[PortfolioService] Error Type: {type(e).__name__}")
            logger.error(f"[PortfolioService] Error Message: {str(e)}")
            logger.error(f"[PortfolioService] Traceback:\n{traceback.format_exc()}")
            raise

File: SERVER/src/main/java/com/ssafy/projectree/domain/porfolio/usecase/PortfolioService.java
package com.ssafy.projectree.domain.porfolio.usecase;

import com.ssafy.projectree.domain.ai.dto.AiPortfolioGenerateDto;
import com.ssafy.projectree.domain.ai.dto.schemas.TaskTechStackSchema;
import com.ssafy.projectree.domain.ai.dto.schemas.UserTaskSchema;
import com.ssafy.projectree.domain.ai.service.InferenceService;
import com.ssafy.projectree.domain.auth.utils.SecurityUtils;
import com.ssafy.projectree.domain.member.model.entity.Member;
import com.ssafy.projectree.domain.node.model.entity.AdvanceNode;
import com.ssafy.projectree.domain.node.model.entity.Node;
import com.ssafy.projectree.domain.node.model.entity.TaskNode;
import com.ssafy.projectree.domain.node.model.repository.NodeRepository;
import com.ssafy.projectree.domain.porfolio.api.dto.PortfolioDto;
import com.ssafy.projectree.domain.porfolio.model.entity.Portfolio;
import com.ssafy.projectree.domain.porfolio.model.repositroy.PortfolioRepository;
import com.ssafy.projectree.domain.tech.model.entity.NodeTechStack;
import com.ssafy.projectree.domain.tech.model.entity.TechStackInfo;
import com.ssafy.projectree.domain.tech.model.entity.TechVocabulary;
import com.ssafy.projectree.domain.tech.model.repository.NodeTechStackRepository;
import com.ssafy.projectree.domain.tech.model.repository.WorkspaceTechStackRepository;
import com.ssafy.projectree.domain.workspace.model.entity.Workspace;
import com.ssafy.projectree.domain.workspace.model.repository.TeamRepository;
import com.ssafy.projectree.domain.workspace.model.repository.WorkspaceRepository;
import com.ssafy.projectree.global.api.code.ErrorCode;
import com.ssafy.projectree.global.exception.BusinessLogicException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
@RequiredArgsConstructor
@Transactional
public class PortfolioService {
    private final WorkspaceRepository workspaceRepository;
    private final PortfolioRepository portfolioRepository;
    private final InferenceService inferenceService;
    private final NodeRepository nodeRepository;
    private final TeamRepository teamRepository;
    private final NodeTechStackRepository nodeTechStackRepository;
    private final WorkspaceTechStackRepository workspaceTechStackRepository;


    @Transactional(readOnly = true)
    public PortfolioDto.Response findPortfolio(Long workspaceId) {
        Member currentMember = SecurityUtils.getCurrentMember();
        Portfolio portfolio = portfolioRepository.findByWorkspaceIdAndMemberId(currentMember.getId(), workspaceId).orElseThrow(() -> new BusinessLogicException(ErrorCode.PORTFOLIO_NOT_FOUND,
                "아직 포트폴리오가 생성되지 않았습니다. 포트폴리오 생성을 먼저 진행해주세요."));
        return PortfolioDto.Response.builder().id(portfolio.getId()).content(portfolio.getContent()).build();
    }

    public PortfolioDto.Response modifyPortfolio(PortfolioDto.Request request) {
        Member currentMember = SecurityUtils.getCurrentMember();
        Portfolio portfolio = portfolioRepository.findById(request.getId()).orElseThrow(() -> new BusinessLogicException(ErrorCode.PORTFOLIO_NOT_FOUND));
        if (!portfolio.getMember().getId().equals(currentMember.getId())) {
            throw new BusinessLogicException(ErrorCode.PORTFOLIO_INVALID_ACCESS_ERROR, "본인만 포트폴리오를 조회 할 수 있습니다.");
        }
        portfolio.setContent(request.getContent());
        Portfolio save = portfolioRepository.save(portfolio);
        return PortfolioDto.Response.builder().id(save.getId()).content(save.getContent()).build();
    }

    public PortfolioDto.Response generatePortfolio(Long workspaceId) {
        Member currentMember = SecurityUtils.getCurrentMember();

        Workspace workspace = workspaceRepository.findById(workspaceId)
                .orElseThrow(() -> new BusinessLogicException(ErrorCode.WORKSPACE_NOT_FOUND));
        List<String> workspaceTechStacks = workspaceTechStackRepository.findTechNamesByWorkspace(workspace);
        int headCount = teamRepository.countByWorkspace(workspace);
        List<TaskNode> taskNodes = nodeRepository.findTaskNodesByWorkspaceAndMember(workspaceId, currentMember.getId());
        List<AdvanceNode> advanceNodes = nodeRepository.findAdvanceNodesByWorkspaceAndMember(workspaceId, currentMember.getId());

        List<UserTaskSchema> taskNodeSchemas = taskNodes.stream().map(this::mapNodeToSchema).toList();
        List<UserTaskSchema> advanceNodeSchemas = advanceNodes.stream().map(this::mapNodeToSchema).toList();

        List<UserTaskSchema> userTaskSchemas = Stream.concat(taskNodeSchemas.stream(), advanceNodeSchemas.stream())
                .collect(Collectors.toList());

        AiPortfolioGenerateDto.Request request = AiPortfolioGenerateDto.Request.builder()
                .projectTitle(workspace.getName())
                .projectStartDate(workspace.getStartDate())
                .projectEndDate(workspace.getEndDate())
                .projectTechStack(workspaceTechStacks)
                .projectDescription(workspace.getDescription())
                .projectHeadCount(headCount)
                .memberId(currentMember.getId())
                .workspaceId(workspace.getId())
                .userTaskSchemas(userTaskSchemas)
                .build();

        AiPortfolioGenerateDto.Response response = inferenceService.generatePortfolio(request);

        Portfolio portfolio = Portfolio.builder()
                .content(response.getContent())
                .member(currentMember)
                .workspace(workspace)
                .build();

        Portfolio savedPortfolio = portfolioRepository.save(portfolio);

        return PortfolioDto.Response.builder()
                .id(savedPortfolio.getId())
                .content(savedPortfolio.getContent())
                .build();
    }

    private UserTaskSchema mapNodeToSchema(Node node) {
        List<NodeTechStack> techStacks = nodeTechStackRepository.findAllByNode(node);
        NodeTechStack selectedTechStack = techStacks.stream()
                .filter(nts -> nts.getTechStackInfo() != null && nts.getTechStackInfo().isSelected())
                .findFirst()
                .orElse(techStacks.isEmpty() ? null : techStacks.get(0));

        TaskTechStackSchema techStackSchema = null;
        if (selectedTechStack != null && selectedTechStack.getTechStackInfo() != null && selectedTechStack.getTechVocabulary() != null) {
            TechStackInfo info = selectedTechStack.getTechStackInfo();
            TechVocabulary vocab = selectedTechStack.getTechVocabulary();
            techStackSchema = TaskTechStackSchema.builder()
                    .name(vocab.getName())
                    .advantage(info.getAdvantage())
                    .disadvantage(info.getDisadvantage())
                    .description(info.getDescription())
                    .build();
        }

        return UserTaskSchema.builder()
                .taskName(node.getName())
                .taskDescription(node.getDescription())
                .taskNote(node.getNote())
                .techStack(techStackSchema)
                .build();
    }
}
